rmi.go

```diff
@@ -0,0 +1,152 @@
+/*
+   Copyright (C) nerdctl authors.
+   Copyright (C) containerd authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/platforms"
+	refdocker "github.com/containerd/containerd/reference/docker"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+)
+
+var rmiCommand = &cli.Command{
+	Name:      "rmi",
+	Usage:     "Remove one or more images",
+	ArgsUsage: "[flags] IMAGE [IMAGE, ...]",
+	Action:    rmiAction,
+}
+
+func rmiAction(clicontext *cli.Context) error {
+	if clicontext.NArg() == 0 {
+		return errors.Errorf("requires at least 1 argument")
+	}
+
+	client, ctx, cancel, err := newClient(clicontext)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	var (
+		imageStore = client.ImageService()
+		cs         = client.ContentStore()
+	)
+
+	imageList, err := imageStore.List(ctx, "")
+	if err != nil {
+		return err
+	}
+
+	var opts []images.DeleteOpt
+	var imageNotFoundError bool
+
+	if clicontext.NArg() >= 1 {
+		img := clicontext.Args().First()
+		named, err := refdocker.ParseDockerRef(img)
+		if err != nil {
+			return err
+		}
+
+		imgFQIN := named.String()
+
+		digests, err := getImageDigests(ctx, cs, imgFQIN, imageList)
+		if err != nil {
+			return errors.Errorf("Error in getting image digests: %v", err)
+		}
+
+		if err := imageStore.Delete(ctx, imgFQIN, opts...); err != nil {
+			fmt.Fprintf(clicontext.App.Writer, "Error: No such image: %s\n", img)
+			imageNotFoundError = true
+		} else {
+			printDigests(clicontext, imgFQIN, digests)
+		}
+	}
+
+	if clicontext.NArg() == 1 {
+		if imageNotFoundError {
+			os.Exit(1)
+		}
+		os.Exit(0)
+	}
+
+	for _, img := range clicontext.Args().Tail() {
```

Comment:
The loop seems redundant.

Can we write like this:
```go
for i, img := range clicontext.Args() {
  doStuff()
  if i == 0 && imageNotFoundError != nil {
    return imageNotFoundError
  }
}
```

Reply:
Are you able to range over `clicontext.Args()`?

I added this to try it out

```
root@vagrant:~/go/src/github.com/nerdctl# git diff
diff --git a/rmi.go b/rmi.go
index 229ee2a..2b80e3c 100644
--- a/rmi.go
+++ b/rmi.go
@@ -64,6 +64,12 @@ func rmiAction(clicontext *cli.Context) error {
        var opts []images.DeleteOpt
        var imageNotFoundError bool

+       for i, img := range clicontext.Args() {
+               fmt.Println(i+" "+img)
+       }
+
+       os.Exit(0)
+
        if clicontext.NArg() >= 1 {
                img := clicontext.Args().First()
                named, err := refdocker.ParseDockerRef(img)
```

and getting the following error on `make`

```
./rmi.go:67:16: cannot range over clicontext.Args() (type "github.com/urfave/cli/v2".Args)
Makefile:39: recipe for target 'nerdctl' failed
make: *** [nerdctl] Error 2
```

Reply:
https://godoc.org/gopkg.in/urfave/cli.v2#Args

Reply:
Maybe `clicontext.Args().Get(i)`? Trying it now ....

Reply:
Args().Slice() returns []string

Reply:
Fixed.